### DISCUSSION TITLE
Support for the Denote package

### DISCUSSION THREAD

**Discussion by pprevos** [emacs-citar/citar#666] [Q&A]
Hi,

Prot is developing the [Denote package](https://github.com/protesilaos/denote) which is a neat note-taking application. The main feature of the package is that the metadata is encoded in the filename. 

Citar, by default uses the citation key as the filename, which does not work for Denote.

I have read the documentation for `citar-notes-sources` and tried a writing function to create a bibliography note in compliance with the Denote file naming convention.

When creating a note, Citar should create a filename in compliance with the denote naming convention. Citar should obviously be able to recognise citation notes.

I would love some help configuring Citar to work with Denote.

**Comment by pprevos** [comment]
Update, I just got something to work to create a note.

This setup create a note from within Citar. Only minor issue is that the citation key cannot contain any underscores.

      (defun denote-citar-file--create-note (key &optional entry)
    "Create a bibliography note through Citar."
    (denote key '("bib")))

     (setq citar-notes-sources
        `((citar-file .
                      ,(list :name "Notes"
                             :category 'file
                             :items #'citar-file--get-notes
                             :hasitems #'citar-file--has-notes
                             :open #'find-file
                             :create #'denote-citar-file--create-note
                             :transform #'file-name-nondirectory))))

**Reply by bdarcus** [reply]
I thought I'd replied to this, but I guess not.

Great to see this! You'll probably need to write some more note functions for it all to work as you expect.

You can also take a look at this for possible ideas.

https://github.com/emacs-citar/citar-org-roam

**Comment by pprevos** [comment]
Hi @bdarcus , I started writing this package :) 

Prot has helped me a lot to get started. It is my first Emacs package so it will be a steep learning curve.

**Reply by bdarcus** [reply]
Cool.

Maybe report back when it's more fully-functional, and I'll add a link to the README here. There will be a few such notes API packages!

**Reply by bdarcus** [reply]
Also, I see your repo. On [this line](https://github.com/pprevos/denote-citar/blob/85245b32523facb891002354dab7c0fee19d23fd/denote-citar.el#L86); don't use `setq`, but instead `citar-register-notes-source`.

https://github.com/emacs-citar/citar-org-roam/blob/a6cfcb8ebe043ef5ca310a0dfb4cc857a0c4fad6/citar-org-roam.el#L198

~And even with that, you wouldn't want to use `setq`; you'd want to use `add-to-list` or `cl-pushnew`.~ I was [wrong here](https://github.com/emacs-citar/citar/pull/661#issuecomment-1193270777); but if you use our convenience function, you'll be good.

**Comment by pprevos** [comment]
Thanks,

This is my very first attempt to write an Emac package - lots to learn. I'll use citar-org-roam as a template for citar-denote.

Denote will use a Org mode keyword `#+reference: citation-key` as link between note and bibliography.

**Reply by bdarcus** [reply]
> This is my very first attempt to write an Emac package ...

Yes; I know ;-)

Citar was my first as well!

**Reply by bdarcus** [reply]
I will add, based on my experience, once you get something fully working and promote it a bit, and people start using it, you will get help along the way, beyond Prot.

But think about how you will encourage and promote that also (pull requests and such). You want to encourage contributions, but also have a clear idea of what you don't accept, and why.

On that note, I vaguely recall seeing @mclearc was interested in denote, and I know he's been working on this own note package.

**Reply by mclearc** [reply]
I've submitted [consult-notes](https://github.com/mclear-tools/consult-notes) to melpa. It works well with denote, especially if you use multiple directories. It would be pretty straightforward to add a bibliography source as well (see the docs on this). But citar-denote looks like a useful package in any case!

**Comment by pprevos** [comment]
Thanks for consult notes. I need to up-skill myself in E Lisp before I can complete denote-citar.
